### From Julia's website:
> Julia is a high-level, high-performance dynamic programming language for technical computing, with <font color='red'>syntax that is familiar to users</font> of other technical computing environments. It provides a sophisticated compiler, distributed parallel execution, numerical accuracy, and an extensive mathematical function library

In [1]:
a = 1+1

2

In [2]:
str = "this is a string"

"this is a string"

In [3]:
println("the value of a is $a")

the value of a is 2


> Julia is a high-level, high-performance dynamic programming language for technical computing, with syntax that is familiar to users of other technical computing environments. It provides a <font color='red'>sophisticated compiler</font>, distributed parallel execution, numerical accuracy, and an extensive mathematical function library

In [19]:
my_square(x) = x^2
@code_llvm(my_square(1))
@code_llvm(my_square(1.0))

my_other_square{T<:Real}(x::T) = x^2

@code_llvm(my_other_square(1))
@code_llvm(my_other_square(1.0))

@time my_other_square(1)
@time my_square(1)


define i64 @julia_my_square_22183(i64) {
top:
  %1 = mul i64 %0, %0
  ret i64 %1
}

define double @julia_my_square_22184(double) {
top:
  %1 = fmul double %0, %0
  ret double %1
}

define i64 @julia_my_other_square_22182(i64) {
top:
  %1 = mul i64 %0, %0
  ret i64 %1
}

define double @julia_my_other_square_22151(double) {
top:
  %1 = fmul double %0, %0
  ret double %1
}
  0.000003 seconds (4 allocations: 160 bytes)
  0.001339 seconds (4 allocations: 160 bytes)


1

### Notes:
* type inference that resulted in calling different functions for different types of inputs. 
    - `mul` and `fmul` in the case above
* arbitrary code isn't necessairly fast in Julia. If speed of code is a concern, Julia code has to be written with performance in mind

> Julia is a high-level, high-performance dynamic programming language for technical computing, with syntax that is familiar to users of other technical computing environments. It provides a sophisticated compiler, <font color='red'> distributed parallel execution</font>, numerical accuracy, and an extensive mathematical function library

In [ ]:
addprocs()

In [39]:
v = collect(1:100)
@parallel (+) for i=1:100
    v[i]
end

@everywhere my_other_square{T<:Real}(x::T) = x^2
pmap(i->my_other_square(i),v)

100-element Array{Any,1}:
     1
     4
     9
    16
    25
    36
    49
    64
    81
   100
   121
   144
   169
     ⋮
  7921
  8100
  8281
  8464
  8649
  8836
  9025
  9216
  9409
  9604
  9801
 10000

> Julia is a high-level, high-performance dynamic programming language for technical computing, with syntax that is familiar to users of other technical computing environments. It provides a sophisticated compiler, distributed parallel execution, <font color='red'>numerical accuracy</font>, and an extensive mathematical function library

In [47]:
x = 2.15*10^12
y = 1.25*10.0^-5
@show x-y
x = BigFloat(2.15*10^12)
y = BigFloat(1.25*10.0^-5)
@show x-y

x - y = 2.15e12
x - y = 2.149999999999999987499999999999999400978299701758800210882327519357204437255859e+12


2.149999999999999987499999999999999400978299701758800210882327519357204437255859e+12

### Notes
* example from [here](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.22.6768&rep=rep1&type=pdf) page 160
* real answer is `r = 2.1499999999999999875 × 10^12`

In [67]:
const C = 2
f(i) = i + C + C

f (generic function with 1 method)

In [68]:
code_llvm(f,(Float64,))
code_llvm(f,(Int64,))


define double @julia_f_22918(double) {
top:
  %1 = fadd double %0, 2.000000e+00
  %2 = fadd double %1, 2.000000e+00
  ret double %2
}

define i64 @julia_f_22919(i64) {
top:
  %1 = add i64 %0, 4
  ret i64 %1
}


In [70]:
g(i) = @fastmath i + C + C
code_llvm(g,(Float64,))
code_llvm(g,(Int64,))


define double @julia_g_22922(double) {
top:
  %1 = fadd fast double %0, 4.000000e+00
  ret double %1
}

define i64 @julia_g_22923(i64) {
top:
  %1 = add i64 %0, 4
  ret i64 %1
}


In [71]:
h(i) = i + (C + C)
code_llvm(h,(Float64,))
code_llvm(h,(Int64,))


define double @julia_h_22924(double) {
top:
  %1 = fadd double %0, 4.000000e+00
  ret double %1
}

define i64 @julia_h_22925(i64) {
top:
  %1 = add i64 %0, 4
  ret i64 %1
}


> Julia is a high-level, high-performance dynamic programming language for technical computing, with syntax that is familiar to users of other technical computing environments. It provides a sophisticated compiler, distributed parallel execution, numerical accuracy, and an <font color='red'>extensive mathematical function library</font>

A good reference from Julia's website: [http://docs.julialang.org/en/stable/stdlib/]